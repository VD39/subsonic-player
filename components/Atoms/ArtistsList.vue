<script setup lang="ts">
defineProps<{
  artists: BaseArtist[];
}>();
</script>

<template>
  <ul class="commaList">
    <li
      v-for="artist in artists"
      :key="artist.id"
      data-test-id="artist-list-item"
    >
      <NuxtLink :to="`/artist/${artist.id}`" class="link">
        {{ artist.name }}
      </NuxtLink>
    </li>
  </ul>
</template>
