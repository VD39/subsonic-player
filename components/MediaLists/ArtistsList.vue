<script setup lang="ts">
import NoMediaMessage from '@/components/NoMediaMessage/NoMediaMessage.vue';
import MediaListWrapper from './MediaListWrapper.vue';
import ArtistItem from './ArtistItem.vue';

defineProps<{
  artists: Artist[];
  rows?: string;
}>();
</script>

<template>
  <MediaListWrapper v-if="artists.length" :row="rows">
    <ArtistItem v-for="artist in artists" :key="artist.id" :artist="artist" />
  </MediaListWrapper>
  <NoMediaMessage v-else icon="PhUsersThree" message="No artists found." />
</template>
