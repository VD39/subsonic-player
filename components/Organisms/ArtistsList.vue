<script setup lang="ts">
import MediaListWrapper from '@/components/Atoms/MediaListWrapper.vue';
import NoMediaMessage from '@/components/Atoms/NoMediaMessage.vue';
import ArtistItem from '@/components/Organisms/ArtistItem.vue';

defineProps<{
  artists: Artist[] | SimilarArtist[];
}>();
</script>

<template>
  <MediaListWrapper v-if="artists.length">
    <ArtistItem v-for="artist in artists" :key="artist.id" :artist />
  </MediaListWrapper>

  <NoMediaMessage
    v-else
    :icon="IMAGE_DEFAULT_BY_TYPE.artist"
    message="No artists found."
  />
</template>
