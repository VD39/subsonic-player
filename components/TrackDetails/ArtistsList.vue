<script setup lang="ts">
defineProps<{
  artists: BaseArtist[];
}>();
</script>

<template>
  <ul :class="[$style.meta, $style.list]">
    <li
      v-for="artist in artists"
      :key="artist.id"
      :class="$style.listItem"
      data-test-id="artist-list-item"
    >
      <NuxtLink :to="`/artist/${artist.id}`">
        {{ artist.name }}
      </NuxtLink>
    </li>
  </ul>
</template>

<style module>
.list {
  @mixin align-center;
}

.listItem {
  &:not(:last-child)::after {
    content: ',\00a0';
  }

  &:nth-last-child(2)::after {
    content: '\00a0&\00a0';
  }
}
</style>
