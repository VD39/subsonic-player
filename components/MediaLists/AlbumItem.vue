<script setup lang="ts">
import ArtistsList from '@/components/TrackDetails/ArtistsList.vue';
import ImageLink from '@/components/Image/ImageLink.vue';

const props = defineProps<{
  album: Album;
}>();

const toLink = `/album/${props.album.id}`;
</script>

<template>
  <article>
    <ImageLink :to="toLink" :title="album.name" :image="album.image" />

    <p class="clamp clamp2">
      <NuxtLink :to="toLink" :title="`Go to album ${album.name}`">
        {{ album.name }}
      </NuxtLink>
    </p>

    <ArtistsList
      v-if="album.artists.length"
      :artists="album.artists"
      class="clamp clamp2"
    />
  </article>
</template>
