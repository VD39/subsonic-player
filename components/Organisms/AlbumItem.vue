<script setup lang="ts">
import ArtistsList from '@/components/Atoms/ArtistsList.vue';
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import SpinningLoader from '@/components/Atoms/SpinningLoader.vue';
import ImageLink from '@/components/Organisms/ImageLink.vue';

const props = defineProps<{
  album: Album;
  hideArtist?: boolean;
}>();

const { playTracks } = useAudioPlayer();
const { getAlbum } = useAlbum();

const loading = ref(false);

async function playAlbumTracks(albumId: string) {
  loading.value = true;

  const selectedAlbum = await getAlbum(albumId);

  if (selectedAlbum) {
    await playTracks(selectedAlbum.tracks);
  }

  loading.value = false;
}

const buttonProps = computed<ButtonProps>(() => ({
  icon: loading.value ? SpinningLoader : ICONS.play,
  text: `Play album ${props.album.name}`,
}));

const toLink = `/album/${props.album.id}`;
</script>

<template>
  <article>
    <div :class="$style.albumImageWrapper">
      <ImageLink
        :to="toLink"
        :title="`Go to album ${album.name}`"
        :image="album.image"
        class="mBS"
      />

      <ButtonLink
        :icon="buttonProps.icon"
        :class="['themeHoverButton', $style.buttonLink]"
        :title="buttonProps.text"
        @click="playAlbumTracks(album.id)"
      >
        {{ buttonProps.text }}
      </ButtonLink>
    </div>

    <p class="mBXS strong smallFont clamp2">
      <NuxtLink
        :to="toLink"
        class="link"
        :aria-label="`Go to album ${album.name}`"
      >
        {{ album.name }}
      </NuxtLink>
    </p>

    <ArtistsList
      v-if="!hideArtist && album.artists.length"
      :artists="album.artists"
      class="smallFont clamp2"
    />
  </article>
</template>

<style module>
.albumImageWrapper {
  position: relative;

  @media (hover: hover) {
    &:hover,
    &:focus {
      .buttonLink {
        --album-play-button-opacity: 1;
      }
    }
  }
}

.buttonLink {
  --album-play-button-opacity: 0;

  position: absolute;
  inset: auto var(--default-space) var(--default-space) auto;
  opacity: var(--album-play-button-opacity);
  transition: opacity var(--transition);
}
</style>
