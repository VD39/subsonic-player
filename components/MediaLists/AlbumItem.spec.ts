import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import { formattedAlbumMock } from '@/test/fixtures';
import ArtistsList from '@/components/TrackDetails/ArtistsList.vue';
import AlbumItem from './AlbumItem.vue';

function factory(props = {}) {
  return mount(AlbumItem, {
    props: {
      album: formattedAlbumMock,
      ...props,
    },
  });
}

describe('AlbumItem', () => {
  let wrapper: VueWrapper;

  beforeAll(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when album artist is not an empty array', () => {
    it('shows the ArtistsList component', () => {
      expect(wrapper.findComponent(ArtistsList).exists()).toBe(true);
    });
  });

  describe('when album artist is an empty array', () => {
    beforeAll(() => {
      wrapper = factory({
        album: {
          ...formattedAlbumMock,
          artists: [],
        },
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('does not show the ArtistsList component', () => {
      expect(wrapper.findComponent(ArtistsList).exists()).toBe(false);
    });
  });
});
