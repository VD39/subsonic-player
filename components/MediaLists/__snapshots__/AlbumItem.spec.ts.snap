// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AlbumItem > matches the snapshot 1`] = `
"<article>
  <figure class="figure"><a title="Go to name">
      <div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="coverArt" alt="name" style="display: none;"></div>
    </a></figure>
  <p class="clamp clamp2"><a title="Go to album name">name</a></p>
  <ul class="commaList clamp clamp2">
    <li data-test-id="artist-list-item"><a>name</a></li>
    <li data-test-id="artist-list-item"><a>name1</a></li>
  </ul>
</article>"
`;

exports[`AlbumItem > when album artist is an empty array > matches the snapshot 1`] = `
"<article>
  <figure class="figure"><a title="Go to name">
      <div class="preloadImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="coverArt" alt="name" style="display: none;"></div>
    </a></figure>
  <p class="clamp clamp2"><a title="Go to album name">name</a></p>
  <!--v-if-->
</article>"
`;
